File: ios/BFExplorer/Plaoc/IPCCore/helper/urlHelper.swift
//
//  urlHelper.swift
//  BFExplorer
//
//  Created by ui03 on 2023/2/14.
//

import UIKit
import SwiftyJSON

let URL_BASE = "http://localhost"

class urlHelper: NSObject {

    
//    static func URL_BASE(url: String) -> String {
//
//        if url.hasPrefix("http:")
//           || url.hasPrefix("https:")
//           || url.hasPrefix("file:")
//           || url.hasPrefix("chrome-extension:") {
//            return url
//        }
//        return "http://localhost"
//    }
    
    static func parseUrl(urlString: Any, base: String = URL_BASE) -> URL? {
        
        guard urlString is String || urlString is URL else { return nil }
        
        var paramURLString = ""
        if urlString is URL {
            let paramURL = urlString as! URL
            paramURLString = paramURL.absoluteString
        }
        
        if urlString is String {
            paramURLString = urlString as! String
        }
        
        guard paramURLString.count > 0 else { return nil }
        
        let baseURL = URL(string: base)
        
        return URL(string: paramURLString, relativeTo: baseURL)?.absoluteURL
        
    }
    
    static func updateUrlOrigin(url: Any, new_origin: String) -> URL? {
        
        let newUrl = urlHelper.parseUrl(urlString: url)
        var urlString = newUrl?.absoluteString
        let origin = newUrl?.absoluteString.analysisURLFormat() ?? ""
        urlString = urlString?.replacingOccurrences(of: origin, with: new_origin)
        return URL(string: urlString ?? "")
    }
    
    static func buildUrl(url: URL, ext: Ext) -> URL? {
        
        var path: String = ""
        var query: String = ""
        
        if ext.pathname != nil {
            path = ext.pathname!
        }
        if (ext.search != nil) {
            if ext.search! is String {
                query = ext.search as? String ?? ""
            } else  if ext.search is [String: Any] {
                var tmpDict = ext.search as! [String: Any]
                for (key, value) in tmpDict {
                    if value is String {
                        
                    } else {
                        tmpDict[key] = ChangeTools.tempAnyToString(value:value)
                    }
                }
                query = ChangeTools.dicValueString(tmpDict) ?? ""
            }
        }
        var components = URLComponents(string: url.absoluteString)
        components?.path = path
        components?.query = query
        
        return components?.url
    }
    
}

struct Ext {
    
    var search: Any?
    var pathname: String?
}

File: next/ts/electron/esm/helper/urlHelper.js
import * as dntShim from "../_dnt.shims.js";
const URL_BASE = "document" in dntShim.dntGlobalThis
    ? document.baseURI
    : "location" in dntShim.dntGlobalThis &&
        (location.protocol === "http:" ||
            location.protocol === "https:" ||
            location.protocol === "file:" ||
            location.protocol === "chrome-extension:")
        ? location.href
        : "file:///";
export const parseUrl = (url, base = URL_BASE) => {
    return new URL(url, base);
};
export const updateUrlOrigin = (url, new_origin) => {
    const { origin, href } = parseUrl(url);
    return new URL(new_origin + href.slice(origin.length));
};
export const buildUrl = (url, ext) => {
    if (ext.pathname !== undefined) {
        url.pathname = ext.pathname;
    }
    if (ext.search) {
        if (ext.search instanceof URLSearchParams) {
            url.search = ext.search.toString();
        }
        else if (typeof ext.search === "string") {
            url.search = ext.search.toString();
        }
        else {
            url.search = new URLSearchParams(Object.entries(ext.search).map(([key, value]) => {
                return [
                    key,
                    typeof value === "string" ? value : JSON.stringify(value),
                ];
            })).toString();
        }
    }
    return url;
};
